File: frontend/src/lib/api/ws.ts
import { writable } from 'svelte/store';
import { getApiBase, getStoredToken } from '$lib/api/client';
import { refreshPlaybackState, refreshPlaybackRuntime } from '$lib/stores/player';
import { handleSyncProgress, handleSyncComplete, handleSyncFailed, loadTidalStatus } from '$lib/stores/tidal';
import { handleTrainingProgress, handleTrainingComplete } from '$lib/stores/training';
import { handleAnalysisProgress, handleAnalysisComplete } from '$lib/stores/audio_analysis';
import { handleAcrCloudProgress, handleAcrCloudComplete } from '$lib/stores/acrcloud';
import { handleDiscoverySpaceRefreshed, setRefreshProgress } from '$lib/components/DiscoverSpace/discover_space_store';
import { setExclusiveEngaged, setExclusiveFailed, setExclusiveReleased } from '$lib/stores/exclusive_status';
import { showToast } from '$lib/stores/toast';

export const wsConnected = writable(false);

export type WsMessage =
	| { type: 'connected' }
	| { type: 'playback_changed' }
	| { type: 'track_changed'; track_id: number }
	| { type: 'queue_updated' }
	| { type: 'listen_history_updated'; track_id: number }
	| { type: 'playback_failed'; message: string }
	| { type: 'library_synced' }
	| { type: 'musicbrainz_enriched' }
	| { type: 'sync_progress'; service: string; progress: number }
	| { type: 'sync_failed'; service: string; message: string }
	| { type: 'training_progress'; stage: string; progress: number; message: string; current_track_id: number | null; current_track_title: string | null; tracks_done: number; tracks_total: number }
	| { type: 'audio_analysis_progress'; analyzed: number; total: number; mode: string }
	| { type: 'audio_analysis_complete'; analyzed: number }
	| { type: 'track_analyzed'; track_id: number }
	| { type: 'acrcloud_scan_progress'; scanned: number; total: number; matches_found: number }
	| { type: 'acrcloud_scan_complete'; scanned: number; matches_found: number }
	| { type: 'discovery_space_refresh_progress'; seed_track_id: number; stage: string; progress: number }
	| { type: 'discovery_space_refreshed'; seed_track_id: number }
	| { type: 'audio_exclusive_engaged'; device: string; transport_format: string }
	| { type: 'audio_exclusive_failed'; device: string; reason: string }
	| { type: 'audio_exclusive_released'; device: string };

export const wsMessages = writable<WsMessage[]>([]);

let socket: WebSocket | null = null;
let reconnectTimer: ReturnType<typeof setTimeout> | undefined;
let queueRefreshTimer: ReturnType<typeof setTimeout> | null = null;

function scheduleQueueRefresh() {
	if (queueRefreshTimer) clearTimeout(queueRefreshTimer);
	queueRefreshTimer = setTimeout(() => {
		queueRefreshTimer = null;
		void refreshPlaybackState();
	}, 100);
}

function getWebSocketUrl(): string {
	const apiUrl = new URL(getApiBase());
	apiUrl.protocol = apiUrl.protocol === 'https:' ? 'wss:' : 'ws:';
	apiUrl.pathname = '/ws';
	apiUrl.hash = '';
	const token = getStoredToken();
	if (token) {
		apiUrl.searchParams.set('token', token);
	}
	return apiUrl.toString();
}

export function connectWebSocket() {
	// Skip if a socket is already up OR currently connecting — avoids creating
	// a second socket while the first is mid-handshake.
	if (socket?.readyState === WebSocket.OPEN || socket?.readyState === WebSocket.CONNECTING) return;

	socket = new WebSocket(getWebSocketUrl());

	socket.onopen = () => {
		wsConnected.set(true);
	};

	socket.onmessage = (event) => {
		try {
			const data = JSON.parse(event.data);
			wsMessages.update((msgs) => [...msgs.slice(-99), data]);
			if (data?.type === 'queue_updated') {
				scheduleQueueRefresh();
			} else if (
				data?.type === 'connected' ||
				data?.type === 'playback_changed' ||
				data?.type === 'track_changed' ||
				data?.type === 'listen_history_updated' ||
				data?.type === 'playback_failed'
			) {
				void refreshPlaybackState();
			}
			if (data?.type === 'track_changed' || data?.type === 'playback_changed') {
				void refreshPlaybackRuntime();
			}
			if (data?.type === 'connected') {
				void loadTidalStatus();
			}
			if (data?.type === 'sync_progress' && data?.service === 'tidal') {
				handleSyncProgress(data.progress ?? 0);
			}
			if (data?.type === 'library_synced') {
				handleSyncComplete();
			}
			if (data?.type === 'sync_failed' && data?.service === 'tidal') {
				handleSyncFailed(data.message ?? 'TIDAL sync failed');
			}
			if (data?.type === 'training_progress') {
				handleTrainingProgress(data);
				// Check if this is the final message (evaluate stage at 96%)
				if (data.stage === 'evaluate' && data.progress >= 0.95) {
					// Give a small delay to let the final state settle
					setTimeout(() => handleTrainingComplete(), 500);
				}
			}
			if (data?.type === 'audio_analysis_progress') {
				handleAnalysisProgress(data);
			}
			if (data?.type === 'audio_analysis_complete') {
				handleAnalysisComplete(data);
			}
			if (data?.type === 'track_analyzed') {
				window.dispatchEvent(
					new CustomEvent('noor:dsp_updated', {
						detail: { trackId: data.track_id }
					})
				);
			}
			if (data?.type === 'acrcloud_scan_progress') {
				handleAcrCloudProgress(data);
			}
			if (data?.type === 'acrcloud_scan_complete') {
				handleAcrCloudComplete(data);
			}
			if (data?.type === 'discovery_space_refresh_progress') {
				setRefreshProgress(data.seed_track_id, data.stage, data.progress);
			}
			if (data?.type === 'discovery_space_refreshed') {
				handleDiscoverySpaceRefreshed(data.seed_track_id);
			}
			if (data?.type === 'audio_exclusive_engaged') {
				setExclusiveEngaged(data.device ?? '', data.transport_format ?? null);
			}
			if (data?.type === 'audio_exclusive_failed') {
				setExclusiveFailed(data.device ?? '', data.reason ?? 'Exclusive mode unavailable');
				// One transient toast on the failure transition so the user notices
				// even if they're not on the settings page. The persistent banner on
				// settings carries the rest of the message.
				showToast(`Exclusive mode unavailable: ${data.reason ?? 'unknown reason'}`, 'error', 8000);
			}
			if (data?.type === 'audio_exclusive_released') {
				setExclusiveReleased(data.device ?? '');
			}
		} catch {}
	};

	socket.onclose = () => {
		wsConnected.set(false);
		// Reconnect after 3s
		reconnectTimer = setTimeout(connectWebSocket, 3000);
	};

	socket.onerror = () => {
		socket?.close();
	};
}

export function disconnectWebSocket() {
	clearTimeout(reconnectTimer);
	if (queueRefreshTimer) {
		clearTimeout(queueRefreshTimer);
		queueRefreshTimer = null;
	}
	socket?.close();
	socket = null;
	wsConnected.set(false);
}

File: noor-server/src/server/ws.rs
use crate::{AppEvent, SharedState};
use axum::{
    Router,
    extract::{
        State, WebSocketUpgrade,
        ws::{Message, WebSocket},
    },
    response::Response,
    routing::get,
};
use serde_json::json;
use tracing::info;

pub fn ws_routes(state: SharedState) -> Router {
    Router::new()
        .route("/ws", get(ws_handler))
        .with_state(state)
}

async fn ws_handler(ws: WebSocketUpgrade, State(state): State<SharedState>) -> Response {
    ws.on_upgrade(|socket| handle_socket(socket, state))
}

async fn handle_socket(mut socket: WebSocket, state: SharedState) {
    info!("WebSocket client connected");

    // Subscribe to the event bus
    let mut rx = {
        let state = state.read().await;
        state.event_tx.subscribe()
    };

    // Send initial state
    let init_msg = json!({
        "type": "connected",
        "message": "Welcome to NOOR"
    });
    if socket
        .send(Message::Text(init_msg.to_string().into()))
        .await
        .is_err()
    {
        return;
    }

    // Forward events to the client
    loop {
        tokio::select! {
            // Events from the app
            recv_result = rx.recv() => {
                let event = match recv_result {
                    Ok(e) => e,
                    // Channel lagged (burst of events exceeded capacity) — re-sync client state.
                    Err(tokio::sync::broadcast::error::RecvError::Lagged(_)) => {
                        let _ = socket.send(Message::Text(json!({"type": "playback_changed"}).to_string().into())).await;
                        continue;
                    }
                    // All senders dropped — server shutting down.
                    Err(tokio::sync::broadcast::error::RecvError::Closed) => break,
                };
                let msg = match event {
                    AppEvent::PlaybackStateChanged => json!({"type": "playback_changed"}),
                    AppEvent::LibrarySynced => json!({"type": "library_synced"}),
                    AppEvent::RadioSimilarityComputed { pairs } => json!({"type": "radio_similarity_computed", "pairs": pairs}),
                    AppEvent::MusicBrainzEnriched => json!({"type": "musicbrainz_enriched"}),
                    AppEvent::TrackChanged { track_id } => json!({"type": "track_changed", "track_id": track_id}),
                    AppEvent::SyncProgress { service, progress } => json!({"type": "sync_progress", "service": service, "progress": progress}),
                    AppEvent::SyncFailed { service, message } => json!({"type": "sync_failed", "service": service, "message": message}),
                    AppEvent::QueueUpdated => json!({"type": "queue_updated"}),
                    AppEvent::ListenHistoryUpdated { track_id } => json!({"type": "listen_history_updated", "track_id": track_id}),
                    AppEvent::PlaybackFailed { message } => json!({"type": "playback_failed", "message": message}),
                    AppEvent::TrainingProgress { stage, progress, message, current_track_id, current_track_title, tracks_done, tracks_total } => json!({
                        "type": "training_progress",
                        "stage": stage,
                        "progress": progress,
                        "message": message,
                        "current_track_id": current_track_id,
                        "current_track_title": current_track_title,
                        "tracks_done": tracks_done,
                        "tracks_total": tracks_total
                    }),
                    AppEvent::AudioAnalysisProgress { analyzed, total, mode } => json!({
                        "type": "audio_analysis_progress",
                        "analyzed": analyzed,
                        "total": total,
                        "mode": mode
                    }),
                    AppEvent::AudioAnalysisComplete { analyzed } => json!({
                        "type": "audio_analysis_complete",
                        "analyzed": analyzed
                    }),
                    AppEvent::TrackAnalyzed { track_id } => json!({
                        "type": "track_analyzed",
                        "track_id": track_id
                    }),
                    AppEvent::AcrCloudScanProgress { scanned, total, matches_found } => json!({
                        "type": "acrcloud_scan_progress",
                        "scanned": scanned,
                        "total": total,
                        "matches_found": matches_found
                    }),
                    AppEvent::AcrCloudScanComplete { scanned, matches_found } => json!({
                        "type": "acrcloud_scan_complete",
                        "scanned": scanned,
                        "matches_found": matches_found
                    }),
                    AppEvent::DiscoverySpaceRefreshProgress { seed_track_id, stage, progress } => json!({
                        "type": "discovery_space_refresh_progress",
                        "seed_track_id": seed_track_id,
                        "stage": stage,
                        "progress": progress,
                    }),
                    AppEvent::DiscoverySpaceRefreshed { seed_track_id } => json!({
                        "type": "discovery_space_refreshed",
                        "seed_track_id": seed_track_id,
                    }),
                    AppEvent::AudioExclusiveEngaged { device, transport_format } => json!({
                        "type": "audio_exclusive_engaged",
                        "device": device,
                        "transport_format": transport_format,
                    }),
                    AppEvent::AudioExclusiveFailed { device, reason } => json!({
                        "type": "audio_exclusive_failed",
                        "device": device,
                        "reason": reason,
                    }),
                    AppEvent::AudioExclusiveReleased { device } => json!({
                        "type": "audio_exclusive_released",
                        "device": device,
                    }),
                    AppEvent::DownloadProgress { done, total, current_title } => json!({
                        "type": "download_progress",
                        "done": done,
                        "total": total,
                        "current_title": current_title,
                    }),
                    AppEvent::DownloadItemDone { track_id, ok, already, path, error } => json!({
                        "type": "download_item_done",
                        "track_id": track_id,
                        "ok": ok,
                        "already": already,
                        "path": path,
                        "error": error,
                    }),
                    AppEvent::DownloadComplete { ok, failed } => json!({
                        "type": "download_complete",
                        "ok": ok,
                        "failed": failed,
                    }),
                };
                if socket.send(Message::Text(msg.to_string().into())).await.is_err() {
                    break;
                }
            }
            // Messages from the client
            client_msg = socket.recv() => {
                match client_msg {
                    Some(Ok(Message::Text(text))) => tracing::debug!("WS client: {}", text),
                    Some(Ok(Message::Pong(_))) => {}
                    Some(Ok(Message::Close(_))) | None => break,
                    _ => {}
                }
            }
        }
    }

    info!("WebSocket client disconnected");
}
